### PR TITLE
Update to JupyterLab 4.0.0a31

### PR DESCRIPTION
**Pull request by jtpio**
Fixes https://github.com/jupyter/notebook/issues/6577

- [x] Update `@jupyterlab` packages
- [x] Check API changes from https://github.com/jupyterlab/jupyterlab/pull/13247
- [x] Check https://github.com/jupyter/notebook/issues/6559
- [x] ~Check notifications: https://github.com/jupyterlab/jupyterlab/pull/12959~ -> **not including it for now**
- [x] Check https://github.com/jupyterlab/jupyterlab/pull/13141
- [x] Check the move of `shared-models`: https://github.com/jupyterlab/jupyterlab/pull/13389
- [x] User service: https://github.com/jupyterlab/jupyterlab/pull/12926 -> Notebook 7 uses the default `ServiceManager`
- [x] File ID: https://github.com/jupyterlab/jupyterlab/pull/13246, https://github.com/jupyterlab/jupyterlab/pull/13370
- [x] https://github.com/jupyterlab/jupyterlab/pull/11962
- [x] Check more changes from https://github.com/jupyterlab/jupyterlab/releases/tag/v4.0.0a31
- [x] Fix UI tests for the Table of Contents:

https://user-images.githubusercontent.com/591645/202727895-d1015e80-1cb0-4282-8598-b7d1992fdc3d.mp4

### CODE REVIEW

**Comment by fcollonval** [comment]
FYI about the notification feature - I have a draft (locally) to extract it to link the view to any widget (not necessarily the one of the status widget as in JupyterLab). I hit a refactoring trouble due to `typestyle`. We definitely should get ride of it (except maybe for `LabIcon` - the use case is quite advanced). Then it will be easy to refactor the code to provide a notification view in `ui-components` that you can anchor on any widget that is adequate for the notebook v7 UI.

File: pyproject.toml [hunk 1]
```diff
@@ -58,7 +58,8 @@ test = [
     "pytest-timeout",
     "pytest-console-scripts",
     "ipykernel",
-    "jupyterlab_server[test]>=2.13,<3",
+    "jupyter_server[test]>=2.0.0rc3,<3",
```
Comment (jtpio) [inline pyproject.toml:61]: This addition seems to be necessary to fix https://github.com/jupyterlab/jupyterlab/issues/13438 (probably related to the pre-release).

**Comment by jtpio** [comment]
> FYI about the notification feature

Nice thanks for the heads-up!

> We definitely should get ride of it 

:+1: 

Happy to track that in a follow-up issue.

**Comment by jtpio** [comment]
> Fix UI tests for the Table of Contents

Thanks @brichet for adding UI tests for the TOC which helped catch this issue! (and fix it with https://github.com/jupyter/notebook/pull/6627/commits/47b0c4fe8ed77d640b4b024337d8cb362f97fa74)

**Pull request merged**